org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/ModelBuilder.scala

```diff
@@ -0,0 +1,257 @@
+  override def key = PackageKey(name)
+}
+
+case class RootNode() extends ContainerNode(null) {
+  override def parent = null
+  override def name = "ROOT"
+  override def displayName = name
+  override def key = null
+  def diff(rn:RootNode):(Iterable[Node],Iterable[Node])={
+    val toUpdate= new MutableList[Node]
+    val toRefresh= new MutableList[Node]
+    def visitContainer(cn:ContainerNode, cn1:ContainerNode):Unit={
+      var children = collection.immutable.ListMap[NodeKey, Node]()
+      var hasNew = false
+
+      cn1.children.foreach(p =>
+        cn.children.get(p._1) match {
+          case Some(n) =>
+             visitNode(n, p._2)
+             children += (p._1-> n)
+           case None =>
+             hasNew =true
+             children += (p._1-> p._2)
+        }
+        )
+        if(children.size != cn.children.size)
+          hasNew = true
+        if(hasNew )
+          toUpdate += cn
+        cn.children = children
+      }
+
+    def visitNode(n:Node, n1:Node)= {
+      if(n <= n1)
+           toRefresh += n
+      n match {
+        case cn:ContainerNode =>
+          n1 match {
+            case cn1:ContainerNode =>
+              visitContainer(cn, cn1)
+            case _ =>
+          }
+        case _ =>
+      }
+    }
+    visitContainer(this, rn)
+    (toUpdate,toRefresh)
+  }
+}
+
+case class MethodKey(name: String, pList: List[List[String]] = List()) extends NodeKey
+case class MethodNode(name: String, parent: ContainerNode, val argTypes: List[List[String]] = List())
+    extends Node(true) with HasModifiers with HasReturnType {
+  override def displayName = name
+  override def key = MethodKey(name, argTypes)
+  override def <= (n:Node):Boolean={
+    val b =super.<=(n)
+    n match {
+      case that:MethodNode =>
+        val rt= returnType
+        returnType=that.returnType
+        b || rt != returnType
+      case _ => b
+    }
+  }
+}
+
+//case class FieldKey(name: String) extends NodeKey
+case class ValNode(name: String, parent: ContainerNode, rt: String)
+    extends Node(true) with HasModifiers with HasReturnType {
+
+  override def displayName = name
+  override def key = MethodKey(name)
+  override def <= (n:Node):Boolean={
+    val b =super.<=(n)
+    n match {
+      case that:ValNode =>
+        val rt= returnType
+        returnType=that.returnType
+        b || rt != returnType
+      case _ => b
+    }
+  }
+}
+
+case class VarNode(name: String, parent: ContainerNode, rt: String)
+    extends Node(true) with HasModifiers with HasReturnType {
+  override def displayName = name
+  override def key = MethodKey(name)
+  override def <= (n:Node):Boolean={
+    val b =super.<=(n)
+    n match {
+      case that:VarNode =>
+        val rt= returnType
+        returnType=that.returnType
+        b || rt != returnType
+      case _ => b
+    }
+  }
+}
+
+
+
+object ModelBuilder extends HasLogger {
+
+  def buildTree(comp: IScalaPresentationCompiler,src:SourceFile):RootNode={
```

Comment:
I can't say that I like the mutable nature of the tree. I would expect a much simpler implementation of this method with an immutable tree.